# The Battle of Neighborhood 

## Introduction
An Indian restaurant chain owner (Client / Stakeholder) is willing to open a restaurant in US, but not able to make a decision in which city. This project helps the stakeholder in identyfing the city where he should open restaurant for a flourshing business.

## Business Problem section
US has 52 states and 300+ plus cities having more that 100K population, project is to evaluate various cities and its Metropolitan Statistical Area with Indian American Population. As latest data is not available for the same will populate the same based on the growth rate as per last census. Post which we will select the city having highest Indian American population where the client can open the restaurant. Will be using the Foursquare data to understand where most of the restaurants are established and select the same neighbourhood to open the restaurant.

### U.S. metropolitan city with largest Asian Indian populations data will be used to select the city where the stakeholder can open an Indian Restaurant

### Import necessary Libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # module to convert an address into␣,→latitude and longitude values
# libraries for displaying images
from IPython.display import Image
from IPython.core.display import HTML
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize
!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library
print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0          conda-forge
    geopy:           

# Additional data that i will be using is of U.S. metropolitan city with largest Asian Indian populations is uploaded in below cell

In [2]:
# The code was removed by Watson Studio for sharing.

Metropolitan_Statistical_Area  Indian_American_Population  Total_population  \
0                       Atlanta                       78980           5268860   
1                        Austin                       23503           1716289   
2                     Baltimore                       32193           2710489   
3                        Boston                       62598           4552402   
4                    Bridgeport                       15439            916829   

   Percentage_of_Total_population  
0                           0.015  
1                           0.014  
2                           0.012  
3                           0.014  
4                           0.017

In [4]:
df_data_2.count()

Metropolitan_Statistical_Area     40
Indian_American_Population        40
Total_population                  40
Percentage_of_Total_population    40
dtype: int64

In [7]:
df_data_2['Indian_American_Population'].max()

526133

## Let us identify which city is having maximum Indians in US

In [11]:
print(df_data_2[df_data_2.Indian_American_Population == df_data_2.Indian_American_Population.max()]) 

   Metropolitan_Statistical_Area  Indian_American_Population  \
22                      New York                      526133   

    Total_population  Percentage_of_Total_population  
22          18897109                           0.028  


## Define Foursquare Credentials and Version details of developer account 

In [12]:
CLIENT_ID = '40ABELL2KJ4QKL1KZSUCEKNLXLI0L32EN1HE3MQWU1ZK5YGV' # your Foursquare ID
CLIENT_SECRET = 'PXKPQQRB0BDRB5TXJHR15S3D2LWAHY0LZ5NA5QLOWQNMCJD4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
import types
import pandas as pd
from botocore.client import Config
import ibm_boto3


So let’s start by converting the selected restaurant’s address to its latitude and longitude coordinates. In order
to define an instance of the geocoder, we need to define a user_agent. We will name our agent
foursquare_agent, as shown below.

In [13]:
address = '102 North End Ave, New York, NY'
geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

40.7149555 -74.0153365


In [14]:
search_query = 'Indian'
LIMIT = 100
radius = 10000
print(search_query + ' .... OK!')

Indian .... OK!


In [15]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=40ABELL2KJ4QKL1KZSUCEKNLXLI0L32EN1HE3MQWU1ZK5YGV&client_secret=PXKPQQRB0BDRB5TXJHR15S3D2LWAHY0LZ5NA5QLOWQNMCJD4&ll=40.7149555,-74.0153365&v=20180605&query=Indian&radius=10000&limit=100'

In [16]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f438f690f769205f8dca706'},
 'response': {'venues': [{'id': '5072fe2eb0ed0fcce2497f30',
    'name': 'Indian Biryani Delights (Cart)',
    'location': {'address': 'Water St',
     'crossStreet': 'btwn Maiden Lane & Pine St',
     'lat': 40.71316451366987,
     'lng': -74.01021890317702,
     'labeledLatLngs': [{'label': 'display',
       'lat': 40.71316451366987,
       'lng': -74.01021890317702}],
     'distance': 475,
     'postalCode': '10005',
     'cc': 'US',
     'city': 'New York',
     'state': 'NY',
     'country': 'United States',
     'formattedAddress': ['Water St (btwn Maiden Lane & Pine St)',
      'New York, NY 10005',
      'United States']},
    'categories': [{'id': '4bf58dd8d48988d1cb941735',
      'name': 'Food Truck',
      'pluralName': 'Food Trucks',
      'shortName': 'Food Truck',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/streetfood_',
       'suffix': '.png'},
      'primary': True}],
    'referra

In [17]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

## Let's explore the first Indian Restaurant in our dataframe.

In [18]:
venues = results['response']['venues'][0]['name']
venues

'Indian Biryani Delights (Cart)'

## In below cells will identify which Postal code the restaurant to be opened by the client

In [19]:
import json 
from pandas.io.json import json_normalize
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories delivery.id  \
0  [{'id': '4bf58dd8d48988d1cb941735', 'name': 'F...         NaN   
1  [{'id': '4bf58dd8d48988d1cb941735', 'name': 'F...         NaN   
2  [{'id': '4bf58dd8d48988d10f941735', 'name': 'I...      275521   
3  [{'id': '4bf58dd8d48988d1cb941735', 'name': 'F...         NaN   
4  [{'id': '52f2ab2ebcbc57f1066b8b1b', 'name': 'S...         NaN   

                delivery.provider.icon.name  \
0                                       NaN   
1                                       NaN   
2  /delivery_provider_seamless_20180129.png   
3                                       NaN   
4                                       NaN   

              delivery.provider.icon.prefix delivery.provider.icon.sizes  \
0                                       NaN                          NaN   
1                                       NaN                          NaN   
2  https://fastly.4sqi.net/img/general/cap/                     [40, 50]   
3                                       NaN                          NaN   
4                                       NaN                          NaN   

  delivery.provider.name                                       delivery.url  \
0                    NaN                                                NaN   
1                    NaN                                                NaN   
2               seamless  https://www.seamless.com/menu/a-saffron-thread...   
3                    NaN                                                NaN   
4                    NaN                                                NaN   

   hasPerk                        id location.address  ...  \
0    False  5072fe2eb0ed0fcce2497f30         Water St  ...   
1    False  5d18e90b96ba5e00232c527f              NaN  ...   
2    False  510aad1ae4b0bfb2583169b9   98 Chambers St  ...   
3    False  573f435b498e03b8bb8582c7     140 Broadway  ...   
4    False  54a4690e498e43a77f503eba    1 Bowling Grn  ...   

                           location.formattedAddress  \
0  [Water St (btwn Maiden Lane & Pine St), New Yo...   
1                [New York, NY 10007, United States]   
2  [98 Chambers St (at Church St), New York, NY 1...   
3  [140 Broadway, New York, NY 10005, United States]   
4  [1 Bowling Grn, New York, NY 10004, United Sta...   

                             location.labeledLatLngs location.lat  \
0  [{'label': 'display', 'lat': 40.71316451366987...    40.713165   
1  [{'label': 'display', 'lat': 40.715357202741, ...    40.715357   
2  [{'label': 'display', 'lat': 40.71482256030567...    40.714823   
3  [{'label': 'display', 'lat': 40.70253964211186...    40.702540   
4  [{'label': 'display', 'lat': 40.704048, 'lng':...    40.704048   

  location.lng  location.neighborhood location.postalCode location.state  \
0   -74.010219                    NaN               10005             NY   
1   -74.011315                    NaN               10007             NY   
2   -74.007689                    NaN               10007             NY   
3   -74.012727                Tribeca               10005             NY   
4   -74.013737                    NaN               10004             NY   

                                                name    referralId  \
0                     Indian Biryani Delights (Cart)  v-1598263650   
1                      Bismilla Indian Biryani House  v-1598263650   
2                      A Saffron Thread Fresh Indian  v-1598263650   
3                         Indian king Biriyani House  v-1598263650   
4  National Museum of the American Indian Museum ...  v-1598263650   

  venuePage.id  
0          NaN  
1          NaN  
2    432263795  
3          NaN  
4          NaN  

[5 rows x 25 columns]

In [20]:
dataframe[["name","location.neighborhood","location.city","location.postalCode","location.labeledLatLngs"]]

name location.neighborhood  \
0                      Indian Biryani Delights (Cart)                   NaN   
1                       Bismilla Indian Biryani House                   NaN   
2                       A Saffron Thread Fresh Indian                   NaN   
3                          Indian king Biriyani House               Tribeca   
4   National Museum of the American Indian Museum ...                   NaN   
5              National Museum of the American Indian                   NaN   
6           Indian Beautiful Art (IBA) Crafts PVT Ltd                   NaN   
7                                      Indian Express                   NaN   
8           Indian Beautiful Art (IBA) Crafts PVT Ltd                   NaN   
9                              Asya Indian Restaurant                   NaN   
10                               Aahar Indian Cuisine                   NaN   
11                                   Baluchi's Indian                   NaN   
12                               Ruchi Indian Cuisine                   NaN   
13                               Indian Biryani House                   NaN   
14                       Indian Grill & Biryani House                   NaN   
15                               Indian Biryani House                   NaN   
16                               smart indian cuisine                   NaN   
17                       Indian Larry Motorcycles NYC                   NaN   
18                                  Indian Restaurant                   NaN   
19                CKGS Indian Visa Application Center                   NaN   
20                                  Indian Food Truck                   NaN   
21                              Indian fast food cart                   NaN   
22                  Panna II Garden Indian Restaurant                   NaN   
23                         Bombay's Indian Restaurant                   NaN   
24                         Shwarma House Indian Grill    Financial District   
25                              Karahi Indian Cuisine                   NaN   
26                                      Swagat Indian                   NaN   
27                             Village Indian Cuisine                   NaN   
28                                Amin Indian Cuisine                   NaN   
29                           Polash Indian restaurant                   NaN   
30                                         Indian Taj                   NaN   
31                             Mughlai Indian Cuisine                   NaN   
32                                 Indian Curry Mahal                   NaN   
33                Moharani Indian Takeaway Restaurant                   NaN   
34                                Deep Indian Kitchen                 NoMad   
35                        Shawarma House Indian Grill                   NaN   
36                               Raj's Indian Kitchen                   NaN   
37                        Crazy Indian Person's Party                   NaN   
38                    American Indian Community House                   NaN   
39                                 Indian People Deli                   NaN   
40                            Minar Indian Restaurant                   NaN   
41                            Indian Grill Food Truck                   NaN   
42                                        Indian Code                   NaN   
43                    Sangam Authentic Indian Cuisine                   NaN   
44                          Yellow Indian Curry Truck                   NaN   
45                         Gandhi Fine Indian Cuisine                   NaN   
46                 Indian Astrologer and Love Psychic                   NaN   
47                        Halal Indian Food & Falafel                   NaN   
48        Indian Psychic Spiritual Healer - Astrology                   NaN   
49                                  Pearl Indian Food                   NaN   

       location.cit

## As Neighborhood is not provided in Foursquare data for all records we will use postal code to shortlist the area where the client has to open his Indian Restaurant

In [21]:
df1 = dataframe[dataframe['location.city'] == 'New York'].reset_index(drop=True)

Let us see few columns of data with Postal code

In [22]:
df1[['name','location.lat','location.lng',"location.city","location.postalCode"]]

name  location.lat  \
0                      Indian Biryani Delights (Cart)     40.713165   
1                       Bismilla Indian Biryani House     40.715357   
2                       A Saffron Thread Fresh Indian     40.714823   
3                          Indian king Biriyani House     40.702540   
4   National Museum of the American Indian Museum ...     40.704048   
5              National Museum of the American Indian     40.704304   
6           Indian Beautiful Art (IBA) Crafts PVT Ltd     40.712281   
7                                      Indian Express     40.713413   
8           Indian Beautiful Art (IBA) Crafts PVT Ltd     40.716135   
9                                Aahar Indian Cuisine     40.713307   
10                                   Baluchi's Indian     40.715672   
11                               Ruchi Indian Cuisine     40.709962   
12                               Indian Biryani House     40.728253   
13                       Indian Grill & Biryani House     40.704245   
14                               Indian Biryani House     40.704955   
15                               smart indian cuisine     40.709022   
16                                  Indian Restaurant     40.719978   
17                CKGS Indian Visa Application Center     40.744725   
18                              Indian fast food cart     40.759469   
19                  Panna II Garden Indian Restaurant     40.726273   
20                         Bombay's Indian Restaurant     40.703569   
21                         Shwarma House Indian Grill     40.703876   
22                              Karahi Indian Cuisine     40.723621   
23                                      Swagat Indian     40.748046   
24                           Polash Indian restaurant     40.799745   
25                             Mughlai Indian Cuisine     40.780206   
26                                 Indian Curry Mahal     40.726337   
27                                Deep Indian Kitchen     40.742025   
28                        Shawarma House Indian Grill     40.701598   
29                    American Indian Community House     40.715691   
30                            Minar Indian Restaurant     40.758199   
31                    Sangam Authentic Indian Cuisine     40.729073   

    location.lng location.city location.postalCode  
0     -74.010219      New York               10005  
1     -74.011315      New York               10007  
2     -74.007689      New York               10007  
3     -74.012727      New York               10005  
4     -74.013737      New York               10004  
5     -74.013759      New York               10004  
6     -74.010129      New York               10007  
7     -74.008469      New York               10007  
8     -74.006610      New York               10007  
9     -74.007994      New York               10007  
10    -74.007652      New York                 NaN  
11    -74.012969      New York               10006  
12    -74.007321      New York               10014  
13    -74.009117      New York               10004  
14    -74.013003      New York                 NaN  
15    -74.009117      New York               10038  
16    -73.995903      New York               10013  
17    -73.996796      New York               10011  
18    -73.972256      New York               10022  
19    -73.986273      New York               10003  
20    -74.011141      New York               10004  
21    -74.008730      New York               10041  
22    -74.003124      New York               10013  
23    -73.993108      New York               10001  
24    -73.938747      New York               10035  
25    -73.950088      New York               10128  
26    -73.988930      New York               10003  
27    -73.990558      New York               10010  
28    -74.011194      New York                 NaN  
29    -73.993469      New York               10002  
30    -73.983938      New York               10036  
31    -74

In [23]:
df1.sort_values(by=['location.postalCode'], inplace=True)

In [24]:
df1['location.postalCode']

23    10001
29    10002
26    10003
19    10003
20    10004
13    10004
4     10004
5     10004
0     10005
3     10005
11    10006
7     10007
6     10007
2     10007
1     10007
8     10007
9     10007
27    10010
17    10011
31    10012
22    10013
16    10013
12    10014
18    10022
24    10035
30    10036
15    10038
21    10041
25    10128
10      NaN
14      NaN
28      NaN
Name: location.postalCode, dtype: object

In [25]:
df1[['location.postalCode']].groupby(['location.postalCode']).size()

location.postalCode
10001    1
10002    1
10003    2
10004    4
10005    2
10006    1
10007    6
10010    1
10011    1
10012    1
10013    2
10014    1
10022    1
10035    1
10036    1
10038    1
10041    1
10128    1
dtype: int64

# Concluding that the Stake holder to open the restaurant in Postalcode 10007 as it has maximum number of Indian restaurants with MSA name as New York-Northern New Jersey-Long Island, NY-NJ-CT-PA